contracts/utils/math/Math.sol

```diff
@@ -114,7 +131,7 @@ library Math {
         // but the largest value we can obtain is type(uint256).max - 1, which happens
         // when a = type(uint256).max and b = 1.
         unchecked {
-            return a == 0 ? 0 : (a - 1) / b + 1;
+            return choice(a == 0, 0, (a - 1) / b + 1);
```

Comment:
looking at this I'm strating to have mixed feelings.

When you do a ternary choice, the compiler can "optimize" it so that only one branch is computer. However, when you use a dedicated function, values for both side must be computed. This remove the opportunity for lazy evaluation.

If both side are very simple math (or at least if the side that is often ignore is cheap enough) then that may make sens ... but if both branches are storage read, then using choice might be a very bad idea.

They may also be cases where the branch that is not needed cannot be computed (overflow, out of array access, ...)

I think we must evaluate gas costs of all the functions where we plan on using choice individually.

Reply:
That's why it should be used only for simple ternary operations, definitively no recommended for nested ternary operations for example, but in this case specifically it is still more optimized than the regular method, once one subtraction, addition and division are cheaper than one jump.

[SUB](https://www.evm.codes/#03?fork=cancun) and [ADD](https://www.evm.codes/#01?fork=cancun) cost `3 gas` units, [DIV](https://www.evm.codes/#04?fork=cancun) cost 5 gas units.

While a single [JUMPI](https://www.evm.codes/#01?fork=cancun) cost `10 gas units`, plus 1 [JUMPDEST](https://www.evm.codes/#5b?fork=cancun), plus 3 gas unit for pushing the PC pointer onto the stack.

Reply:
I'm with you on all this.

Reply:
That raises the issue of input sampling in (fuzzing) tests. I believe 0 and 1 and over represent compared to "real" use cases ... but again it's hard to figure out what the real use cases are.

Reply:
IMO by default should be assumed an uniform distribution, it means that no inputs are special, zero is not more likely to be used than `7014971116814175` for example:
```solidity
if (a == 0) return 0;
```
Is no better than doing this:
```solidity
if (a == 7014971116814175) return 83755424;
```
This means that optimizations and branching must be guided on reduce the average gas cost of all inputs.

Reply:
Actually here there's an even better way to optimize this code using branchless method, this method works for any `cond ? a : b` when `a == 0 || b == 0`:
```solidity
unchecked {
    return SafeCast.toUint(a > 0) * ((a - 1) / (b + 1));
}
```

Reply:
Done

This consumes `372 gas` when `a !=0` and  `324` when `a == 0`
```solidity
contract CeilDiv1 {
    function ceilDiv(uint256 a, uint256 b) external pure returns (uint256) {
        if (b == 0) {
            Panic.panic(Panic.DIVISION_BY_ZERO);
        }
        unchecked {
            return a == 0 ? 0 : (a - 1) / b + 1;
        }
    }
}
```

This consumes `355 gas` regardless the value of `a`.
```
contract BranchlessCeilDiv {
    function ceilDiv(uint256 a, uint256 b) external pure returns (uint256) {
        if (b == 0) {
            Panic.panic(Panic.DIVISION_BY_ZERO);
        }
        unchecked {
            return SafeCast.toUint(a > 0) * ((a - 1) / b + 1);
        }
    }
}
```

For simple code like this, this is how I compare the gas without fuzzing:
```python
total = 2 ** 256 // number of possible inputs

// a == 0
// gas cost multiplied by how many inputs can reach this branch.
branch1_weight = 324 * 1

// a != 0
branch2_weight = 374 * (total - 1)

avg_gas_cost = (branch1_weight + branch2_weight) / 2
```

Reply:
IMO by default should be assumed an uniform distribution

I would dissagree with that. I'm readdy to bet that, considering all the number manipualted onchain trough oracles, tokens, NFTs, ... there is significantly more between `0` and `2**255 -1` than between `2**255-1` and `2**256 - 1`.

Most practical values are probably < 10**36

Reply:
> Most practical values are probably < 10**36

If this is the case, and you know the distribution, just need plug the weight of the inputs, example.

Given all inputs `x`:
- `x == 0` weight is `1000` _(this basically means x==0 happens 100x more often than  `0 < x < 10**36`_
- `0 < x < 10**36` weight is `10`
- `x >= 10**36` weight is `1`

Then we do the same weighted gas calculation, with a more scientific way to decide which implementation is more efficient.
The formula is sum of all `input_count * weight * gas_cost` divided by the total
```python
input_1 = 1 # x == 0
input_2 = (2 ** 36) - input_0 # x > 0
input_3 = (2 ** 256) - input_1 - input_0 # x > (2 ** 36)

branch1_weight = input_0 * 1000
branch2_weight = (input_1 * 10) + (input_2 * 1)

total = branch1_weight + branch2_weight

branch1_gas_cost = branch1_weight * 304
branch2_gas_cost = branch2_weight * 350

avg_gas_cost = (branch1_gas_cost + branch2_gas_cost) / total
```
Using this formula, the final gas cost is:
`349` for `a == 0 ? 0 : (a - 1) / b + 1`
`333` for  `SafeCast.toUint(a != 0) * ((a - 1) / b + 1);`

Reply:
Also big values are more common when using decimal arithmetic, for example calculate sqrt with 10 decimal places precision:
```solidity
uint256 internal constant DECIMALS = 10; // precision in decimals

function sqrt(uint256 x) public pure returns (uint256 result) {
    x *= 10 ** (DECIMALS * 2);
    result = Math.sqrt(x);
}

function divCeil(uint256 x) public pure returns (uint256 result) {
    x *= 10 ** DECIMALS;
    result = Math.divCeil(x);
}
```

Reply:
> we don't even know the input distribution

This is the major source of uncertainty when deciding to implement gas optimizations. Most of the times we try to estimate what's better, and I agree that for most practical purposes is < 10 ** 36. This opinion is based exclusively on how popular 1e18 decimals are.

> Also big values are common when using decimal arithmetic

This is a very interesting observation. I tend to agree with this concern but I also have 0 evidence of a sqrt operation being widely used in production with big values, I'd appreciate references if you have.

Considering Uniswap's Universal Router is the #1 mainnet gas consumer, I tried searching for `sqrt` operations and only found sqrt for large values when [minting liquidity in V2](https://github.com/Uniswap/v2-core/blob/ee547b17853e71ed4e0101ccfd52e70d5acded58/contracts/UniswapV2Pair.sol#L95) and on the [TWAMM math of v4](https://github.com/Uniswap/v4-periphery/blob/6616b12db25257ffb3c562f131612ebb2fd89082/contracts/libraries/TWAMM/TwammMath.sol#L44C14-L44C32). I don't consider none of these cases to be a big deal.

Reply:
In this case specifically, regardless of the distribution doing, doing this:
```solidity
// Consumes 333 gas units
SafeCast.toUint(a > 0) * ((a - 1) / b + 1);
```
is cheaper, but maybe less clear, than doing this:
```solidity
// Consumes 302 gas when a == 0, and 350 when a != 0
a == 0 ? 0 : (a - 1) / b + 1;
```

Reply:
In this case specifically, regardless of the distribution, doing this:
```solidity
// Consumes 333 gas units
SafeCast.toUint(a > 0) * ((a - 1) / b + 1);
```
is cheaper, but maybe less clear, than doing this:
```solidity
// Consumes 302 gas when a == 0, and 350 when a != 0
a == 0 ? 0 : (a - 1) / b + 1;
```

Reply:
Yes. Agree with this one specifically.

More generally, I don't think we have enough data to even estimate the total economic cost of this. The savings are 17 gas in this case (can be extrapolated to others) and will require ~10,000 operations to even cost 51 USD on mainnet (at 100 gwei and ETH trading at 3000).

Reply:
Honestly we are only discussing this because most of those optimizations, and actually most of things that `Math.sol` does should be handled by Solidity, not by a library, because most optimization depends on the context where it is being used.
Example calculate the log2 of an `uint128` is cheaper than calculate of an `uint256`, but consider this:
```solidity
uint256 result = Math.log2(Math.sqrt(uint256(value)));
```
The sqrt of an `uint256` cannot return a value greater than `uint128`, sqrt of `uint128` returns an `uint64`, etc... but openzepellin doesn't know about that, but the compiler does.. that's why I think those mathematical function should be built-in the compiler, same way as overflow check, not by a library.

Reply:
As result this cost increases exponentially, so 13 gas unit here in a simple operation, become 100x in more complex formulas.